rfc/store-deliver.md

```diff
@@ -0,0 +1,310 @@
+
+While not relevant for verifiability, it is also important pointing out that the service MAY need to run some data computation, issue billing events or write indexes for retrievability asynchronously. Today, web3.storage operated store service relies on "S3 Bucket events" to achieve this, but it should not be assumed that all stores web3.storage will rely on will provide such features.
+
+## New capabilities
+
+### `store/deliver` capability
+
+After the agent handled the asynchronous task of posting the bytes, the agent can invoke `store/deliver` to notify the service that the bytes were delivered. This invocation in the future MAY include a challenge for the service based on the posted data.
+
+```json
+{
+  "can": "store/deliver",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "url": "https://..."
+  }
+}
+```
+
+### `store/confirm` capability
+
+Wwhen the service handles the invocation of `store/deliver`, and potentially, after some asynchronous computations/verifications, the service invokes `store/confirm` in order to issue a signed receipt that the flow is now finished.
+
+```json
+{
+  "can": "store/confirm",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "url": "https://..."
+  }
+}
+```
+
+## Common flows
+
+Aiming to introduce verifiability into content storing, as well as to pave the way for a future proof of delivery, `store/deliver` capability is proposed. It would also make services not tied with store put events to execute side effects.
+
+To offer verifiability, `store/add` invocation should be associated with `store/deliver` and `store/confirm` via [effect]s. `store/deliver` would be invoked by the client to notify the service that content was stored, and the the service should also sign a confirmation that it is true via `store/confirm`.
+
+### `store/deliver` of not previously delivered content
+
+Taking into account the above, we can look at the following flow walkthrough and diagram:
+
+1. client invokes `store/add`
+
+```json
+{
+  "can": "store/add",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "size": 1234
+  }
+}
+```
+
+2. service issues `store/add` receipt with presigned URL, effect for `store/deliver` task that MAY be invoked in the future by the agent and `store/confirm` tasks that MAY be invoked in the future by the service.
+
+```json
+{
+  "ran": "bafy...storeAdd",
+  "out": {
+    "ok": {
+      "status": "upload",
+      "url": "https://...",
+      "link": "bag...",
+      "allocated": 1234
+    }
+  },
+  "fx": {
+    "join": { "/": "bafy...storeConfirm" },
+    "fork": [{ "/": "bafy...storeDeliver" }]
+  },
+}
+```
+
+3. client posts the content to the provided presigned URL.
+
+4. client invokes `store/deliver`
+
+```json
+{
+  "can": "store/deliver",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "url": "https://..."
```

Comment:
What is the URL here, pre-signed PUT url we handed prior ?

Reply:
yeah, I think it makes sense to be the same, rather than user deriving something. What do you think?

Reply:
Are we using that URL for anything ? Otherwise I would just leave it out.

Reply:
My main concern is a transition period where we change bucket and there were a few on going ops, where then this invocation is received after the switch. This would mean `store/add` handler returned presigned url for bucket A, then we swap to bucket B our back end, then `store/deliver` comes in and we do not know where to claim things. I would therefore like to keep this, unless there are other suggestions